### PR TITLE
feat(Indicator): add new Indicator component to component library

### PR DESCRIPTION
**Pull request by YossiSaadi**
https://monday.monday.com/boards/3532714909/pulses/4340690959

image

### CODE REVIEW

File: src/components/Badge/Badge.module.scss
Comment (hadasfa) [inline src/components/Badge/Badge.module.scss:23]: you can write this declaration once for all positions

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:23]: I think the component is redundant => can be only badge

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:25]: why boolean?

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [reply src/components/Badge/Badge.tsx:23]: Also in the end we decided that the badge will be not a component it will be a type enum (counter or indication)

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:27]: Why? if the badge is hidden there is no need to use this component

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:28]: let's talk about it?

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:37]: let's talk about it?

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:24]: let's use the same position methodology we use in other components (consistency)

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:59]: the class name purpose is always positioning and layouting the component and therefore it should be on the wrapper element.

Comment (hadasfa) [reply src/components/Badge/Badge.tsx:59]: Also I think it will make more sense that the data test id will be on the wrapper element but let's talk about it

File: src/components/Badge/Badge.tsx
Comment (shlomitc) [inline src/components/Badge/Badge.tsx:31]: why don't we call in `NotificationBadge`?

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:17]: we can remove the notes :)

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:23]: I think it can move to Feedback category

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:43]:
```suggestion
Badge component is responsible for positioning an indicator/counter on a child component
```

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:48]: Overview story should be written in a way users will be able to use controls panel, see how we did it in other components :)

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:59]: let's talk about it

File: src/components/Badge/__tests__/badge-snapshot-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-snapshot-tests.jest.tsx:10]: children isn't mandetory?

File: src/components/Badge/__tests__/badge-snapshot-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-snapshot-tests.jest.tsx:48]: gap and circular should be not related tests, no?

File: src/components/Badge/__tests__/badge-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-tests.jest.tsx:21]: If we have proper snapshots there is no need in this test

File: src/components/Badge/__tests__/badge-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-tests.jest.tsx:27]: If we have proper snapshots there is no need in this test

File: src/components/Badge/__tests__/badge-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-tests.jest.tsx:33]: If we have proper snapshots there is no need in this test + children should be mandatory

File: src/components/Badge/__tests__/badge-tests.jest.tsx
Comment (hadasfa) [inline src/components/Badge/__tests__/badge-tests.jest.tsx:41]: If we have proper snapshots there is no need in this test

File: src/components/Indicator/Indicator.module.scss
Comment (hadasfa) [inline src/components/Indicator/Indicator.module.scss:10]: we can convert it to primary instead of colorPrimary and it will be easier to use in the tsx code without using helper funtions

File: src/components/Indicator/Indicator.tsx
Comment (hadasfa) [inline src/components/Indicator/Indicator.tsx:13]: indicator is redundant, so let's switch indicatorColor to color?

File: src/components/Indicator/Indicator.tsx
Comment (hadasfa) [inline src/components/Indicator/Indicator.tsx:14]: sure is needed? worth to see how people approach it in different places/ ask Talya

File: src/components/Indicator/__tests__/indicator-tests.jest.tsx
Comment (hadasfa) [inline src/components/Indicator/__tests__/indicator-tests.jest.tsx:17]: if we have proper snapshots there is no need in this test :)

File: src/components/Indicator/__tests__/indicator-tests.jest.tsx
Comment (hadasfa) [inline src/components/Indicator/__tests__/indicator-tests.jest.tsx:22]: if we have proper snapshots there is no need in this test :)

File: src/components/Indicator/__tests__/indicator-tests.jest.tsx
Comment (hadasfa) [inline src/components/Indicator/__tests__/indicator-tests.jest.tsx:39]: if we have proper snapshots there is no need in this test :)

Comment (hadasfa) [reply src/components/Indicator/__tests__/indicator-tests.jest.tsx:39]: let's add a snapshot for data test id if we don't have one

File: src/components/Indicator/__tests__/indicator-tests.jest.tsx
Comment (hadasfa) [inline src/components/Indicator/__tests__/indicator-tests.jest.tsx:44]: let's add snapshot test for this if we don't have one

**Review by hadasfa — CHANGES_REQUESTED** [review]
Super awesome in general!  most comments about out standards and guidelines, but very good :)

File: src/components/index.js
Comment (shlomitc) [inline src/components/index.js:94]: why do we actually extract this into two different components?

Comment (shlomitc) [inline src/components/index.js:94]: why do we actually need two components here?

**Review by shlomitc — CHANGES_REQUESTED** [review]

File: src/components/Badge/Badge.tsx
Comment (YossiSaadi) [reply src/components/Badge/Badge.tsx:25]: check with Yael if we need another prop for actual px gap
change `gap` naming to something else: `inbound`? `outbound`? think of something (contact Yael also about this)

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:23]: change prop name to type
Also in the end we decided that the badge will be not a component it will be a type enum (counter or indication)

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [reply src/components/Badge/Badge.tsx:31]: In my opinon NotificationBadge is too long and badge is clear, also design request this name :)

File: src/components/index.js
Comment (hadasfa) [reply src/components/index.js:94]: +1

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (YossiSaadi) [reply src/components/Badge/__stories__/Badge.stories.mdx:59]: Should discuss with Yael

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:33]: can you give me context why we add this?

File: src/components/Badge/Badge.tsx
Comment (hadasfa) [inline src/components/Badge/Badge.tsx:34]: cool!

File: src/components/Badge/Indicator/Indicator.module.scss
Comment (hadasfa) [inline src/components/Badge/Indicator/Indicator.module.scss:10]: Let's pass the right color enum to the counter / indiction component instead of override the css (override component styles by custom class names is less recommended practice, I can explain more about it if you want)

File: src/components/Badge/Indicator/Indicator.tsx
Comment (hadasfa) [inline src/components/Badge/Indicator/Indicator.tsx:14]: use our regular function for calculate data test id or not set data test id at all :)

File: src/components/Badge/Badge.tsx
Comment (YossiSaadi) [reply src/components/Badge/Badge.tsx:33]: talked about it in the story meeting

File: src/components/Badge/Indicator/Indicator.tsx
Comment (YossiSaadi) [reply src/components/Badge/Indicator/Indicator.tsx:14]: Since I removed VibeComponentProps (as the component is not "exported") there's no id, so I didn't see a reason for using the getTestId
But on second thought, there's no reason not allowing passing VibeComponentProps to the element

File: src/components/Badge/Indicator/Indicator.module.scss
Comment (YossiSaadi) [reply src/components/Badge/Indicator/Indicator.module.scss:10]: not sure I got ya, can you please elaborate?

Comment (YossiSaadi) [reply src/components/Badge/Indicator/Indicator.module.scss:10]: talked about it :)

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (hadasfa) [inline src/components/Badge/__stories__/Badge.stories.mdx:87]: not sure inline is clear but if Yael ok with it I'm also ok with it :)

**Review by hadasfa — DISMISSED** [review]
Looks good to me, let's see to story live in our sync before merge (and check my notes :) )

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (YossiSaadi) [reply src/components/Badge/__stories__/Badge.stories.mdx:13]: `createStoryMetaSettings` is not exposed anymore since @SergeyRoyt created `vibe-storybook-components`
@SergeyRoyt, right?

**Comment by YossiSaadi** [comment]
> Looks good to me, let's see to story live in our sync before merge (and check my notes :) )

File: webpack/published-ts-components.js
Comment (shlomitc) [inline webpack/published-ts-components.js:4]: name-wize, I think that we wanted to call it `NotifiationBadge`.
I agree that it is generic enough to be called `Badge`, but a `Badge` can be also textual.
we might even go with `BulletBadge`,

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (shlomitc) [inline src/components/Badge/__stories__/Badge.stories.mdx:67]: Since it's adding a counter number by default, it might be better to have this example with the `counterProps`.
image

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (shlomitc) [inline src/components/Badge/__stories__/Badge.stories.mdx:179]: I actually wonder why this isn't the default. it looks much better even in different colors

with border
image

without
image

File: src/components/Badge/Badge.tsx
Comment (shlomitc) [inline src/components/Badge/Badge.tsx:34]: shouldn't it be `counterProps`?

**Review by shlomitc — CHANGES_REQUESTED** [review]
looks really good,

File: webpack/published-ts-components.js
Comment (YossiSaadi) [reply webpack/published-ts-components.js:4]: Actually the `Badge` is the component that responsible for "layouting" types of badges on a child component, so basically we can attach whatever we want to it, even if it is textual
You want to re-discuss the naming in a wider forum? since Yael has decided to go with Badge and I don't want to change something without her knowing

File: src/components/Badge/Badge.tsx
Comment (YossiSaadi) [reply src/components/Badge/Badge.tsx:34]: it should receive any of the `type` option props. 
For now, it's either Counter or Indicator, I guess next we might have more (icons?).
So it should allow to inject props for any badge, to the component itself

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (YossiSaadi) [reply src/components/Badge/__stories__/Badge.stories.mdx:179]: I actually agree on that :)
I'll take it with Yael or Meirav

File: src/components/Badge/Badge.tsx
Comment (shlomitc) [reply src/components/Badge/Badge.tsx:34]: as a consumer of this component, I'd expect on of the following.
1. Use the `Badge` component and pass it direct props like `border`. so if one of its capabilities is displaying a counter, then I'd expect a prop called `counter` to be passed to it.
2. Use the `Badge` component and understand that behind the scenes it uses the `Counter` component, and thus I shall pass the `counterProps` so it will propagate.
3. Use the `Badge` component and compose `Counter` on my end, and then I know exactly which props to pass into the counter component.

The current state is a bit of a mish-mash, so I'd suggest to go either with option 1 or option 2

Comment (YossiSaadi) [reply src/components/Badge/Badge.tsx]: @shlomitc Thanks to your suggestions I was able to achieve pretty solid type checking!
Please check the last commit and tell me what you think.

If user does not pass type of Counter, TS will allow only Indicator props (flat into regular props, not another object), if the user does pass type Counter, the user is able to only type in Counter props (again, flat).
Examples:

image
image
image

Comment (YossiSaadi) [reply src/components/Badge/Badge.tsx]: of course you can also pass other Counter/Indicator props inline (flat)
e.g.:
```
<Badge type={Badge.types.COUNTER} count={100} maxDigits={2}>
  <Button leftIcon={WhatsNew}>What's new</Button>
</Badge>
```

Comment (shlomitc) [reply src/components/Badge/Badge.tsx]: My personal preference is that components will be as dumb as possible, so this flattening is great already.

File: src/components/Badge/__stories__/Badge.stories.mdx
Comment (SergeyRoyt) [reply src/components/Badge/__stories__/Badge.stories.mdx:14]: That's the correct usage, yeah

**Comment by YossiSaadi** [comment]
@hadasfa @shlomitc after a feedback from Meirav and Shlomi - Yael and I did some heavy changes to the component and therefore to the api
So I had to dismiss your previous approvals
sorry..

File: src/components/Badge/BadgeConstants.ts
Comment (YossiSaadi) [inline src/components/Badge/BadgeConstants.ts:19]: Please share your thoughts on the naming, after a brainstorm with Yael this are the best we could come up with

**Comment by YossiSaadi** [comment]
Yael has more changes she wants to apply, putting this on hold

**Review by hadasfa — APPROVED** [review]
I preferred the previous approach but this is also ok

File: src/components/Badge/BadgeConstants.ts
Comment (hadasfa) [reply src/components/Badge/BadgeConstants.ts:19]: I really prefer the previous approach. Maybe round, rectangle and transparent?

**Pull request merged**
